File: src/ins/filters.py
"""
惯性导航系统 (INS) 滤波器与检测器。

包含低通滤波器、垂直通道卡尔曼滤波器和零速检测器。
"""

from collections import deque

import numpy as np


class LowPassFilter:
    """一阶指数移动平均低通滤波器。

    alpha 越小，截止频率越低（平滑效果越强）。
    """

    def __init__(self, alpha=0.05):
        self.alpha = alpha
        self._value = None

    def reset(self):
        self._value = None

    def update(self, raw):
        if self._value is None:
            self._value = float(raw)
        else:
            self._value = self.alpha * raw + (1.0 - self.alpha) * self._value
        return self._value

    @property
    def value(self):
        return self._value


class VerticalKalmanFilter:
    """垂直通道卡尔曼滤波器。

    状态向量 x = [h, v, a]^T
        h — 高度 (m)
        v — 垂直速度 (m/s)
        a — 垂直加速度 (m/s^2)

    预测模型为牛顿运动学（匀加速假设）。
    观测量为气压计高度变化量 z = h + noise。
    """

    def __init__(self, R=0.5, sigma_a=0.5):
        self.x = np.zeros(3)
        self.P = np.eye(3)
        self.R = float(R)
        self.sigma_a = float(sigma_a)

    def reset(self):
        self.x = np.zeros(3)
        self.P = np.eye(3)

    def predict(self, dt, a_imu_z):
        """执行预测步，用 IMU 垂直加速度注入加速度状态分量。"""
        dt2 = dt * dt
        F = np.array([
            [1.0, dt, 0.5 * dt2],
            [0.0, 1.0, dt],
            [0.0, 0.0, 1.0],
        ])
        G = np.array([0.5 * dt2, dt, 1.0])
        Q = (self.sigma_a ** 2) * np.outer(G, G)

        self.x = F @ self.x
        self.x[2] = a_imu_z
        self.P = F @ self.P @ F.T + Q

    def update(self, z_altitude):
        """用气压高度观测值执行更新步。"""
        H = np.array([[1.0, 0.0, 0.0]])
        S = (H @ self.P @ H.T)[0, 0] + self.R
        K = (self.P @ H.T) / S
        y = z_altitude - self.x[0]
        self.x = self.x + (K * y).flatten()
        self.P = (np.eye(3) - K @ H) @ self.P

    def apply_zupt(self):
        """零速更新：将速度状态强制为零，并收缩对应协方差。"""
        self.x[1] = 0.0
        self.P[1, :] *= 0.01
        self.P[:, 1] *= 0.01
        self.P[1, 1] = 1e-4


class ZUPTDetector:
    """零速检测器。

    在滑动窗口内计算加速度模平方的方差和陀螺仪信号模的方差，
    当两者均低于阈值时判定为静止。
    """

    def __init__(self, acc_threshold=0.1, gyro_threshold=0.01, window_size=20):
        self.acc_threshold = acc_threshold
        self.gyro_threshold = gyro_threshold
        self.acc_buffer = deque(maxlen=window_size)
        self.gyro_buffer = deque(maxlen=window_size)

    def reset(self):
        self.acc_buffer.clear()
        self.gyro_buffer.clear()

    def update(self, acc_norm_sq, gyro_norm):
        """添加一个样本并返回是否处于静止状态。

        Parameters
        ----------
        acc_norm_sq : float
            原始加速度三轴平方和 (ax^2 + ay^2 + az^2)。
        gyro_norm : float
            陀螺仪三轴模 sqrt(gx^2 + gy^2 + gz^2)。

        Returns
        -------
        bool
            True 表示检测到静止。
        """
        self.acc_buffer.append(acc_norm_sq)
        self.gyro_buffer.append(gyro_norm)
        if len(self.acc_buffer) < self.acc_buffer.maxlen:
            return False
        acc_var = float(np.var(self.acc_buffer))
        gyro_var = float(np.var(self.gyro_buffer))
        return acc_var < self.acc_threshold and gyro_var < self.gyro_threshold

File: src/ins/Filters.cpp
#include "Filters.h"
#include <cmath>
#include <numeric>
#include <algorithm>

// ============================================================
// Filters.cpp — INS 滤波器实现
// ============================================================

// ── LowPassFilter ─────────────────────────────────────────────

// 构造：保存滤波系数
LowPassFilter::LowPassFilter(double alpha)
    : m_alpha(alpha)
{}

// 重置内部状态（下次 update 将重新初始化）
void LowPassFilter::reset() {
    m_value.reset();
}

// 一阶 IIR 低通滤波：y[n] = alpha * x[n] + (1-alpha) * y[n-1]
// 首次调用直接以原始值初始化
double LowPassFilter::update(double raw) {
    if (!m_value.has_value()) {
        m_value = raw;
    } else {
        m_value = m_alpha * raw + (1.0 - m_alpha) * m_value.value();
    }
    return m_value.value();
}

// 返回当前滤波值（未初始化时为 nullopt）
std::optional<double> LowPassFilter::value() const {
    return m_value;
}

// ── VerticalKalmanFilter ──────────────────────────────────────

// 构造：初始化噪声参数并重置状态
VerticalKalmanFilter::VerticalKalmanFilter(double R, double sigma_a)
    : m_R(R), m_sigmaA(sigma_a)
{
    reset();
}

// 重置为零初始状态：x=[0,0,0]，P=I
void VerticalKalmanFilter::reset() {
    m_x.fill(0.0);
    m_P.fill(0.0);
    // P = I（单位矩阵）
    P(0,0) = 1.0; P(1,1) = 1.0; P(2,2) = 1.0;
}

// 预测步：牛顿运动学状态转移 + 过程噪声
// F = [[1, dt, 0.5*dt²], [0, 1, dt], [0, 0, 1]]
// G = [0.5*dt², dt, 1]（噪声驱动向量）
// Q = sigma_a² * G*G^T
void VerticalKalmanFilter::predict(double dt, double aImuZ) {
    double dt2 = dt * dt;
    // F 矩阵元素（仅非零非对角线元素）
    double F01 = dt, F02 = 0.5*dt2, F12 = dt;

    // 状态转移：x_new = F * x
    double h = m_x[0], v = m_x[1], a = m_x[2];
    m_x[0] = h + F01*v + F02*a;
    m_x[1] = v + F12*a;
    m_x[2] = aImuZ;  // 直接注入 IMU 加速度作为第三状态

    // 过程噪声矩阵 Q = sigma² * G*G^T
    double sa2 = m_sigmaA * m_sigmaA;
    double g0 = 0.5*dt2, g1 = dt, g2 = 1.0;
    double Q[3][3] = {
        {sa2*g0*g0, sa2*g0*g1, sa2*g0*g2},
        {sa2*g1*g0, sa2*g1*g1, sa2*g1*g2},
        {sa2*g2*g0, sa2*g2*g1, sa2*g2*g2},
    };

    // P_new = F * P * F^T + Q（手动展开 3×3 矩阵乘法）
    double Pold[3][3];
    for (int i = 0; i < 3; ++i)
        for (int j = 0; j < 3; ++j)
            Pold[i][j] = P(i, j);

    // F * P（F 是稀疏矩阵：只有对角线和上三角部分）
    double FP[3][3];
    FP[0][0] = Pold[0][0] + F01*Pold[1][0] + F02*Pold[2][0];
    FP[0][1] = Pold[0][1] + F01*Pold[1][1] + F02*Pold[2][1];
    FP[0][2] = Pold[0][2] + F01*Pold[1][2] + F02*Pold[2][2];
    FP[1][0] = Pold[1][0] + F12*Pold[2][0];
    FP[1][1] = Pold[1][1] + F12*Pold[2][1];
    FP[1][2] = Pold[1][2] + F12*Pold[2][2];
    FP[2][0] = Pold[2][0];
    FP[2][1] = Pold[2][1];
    FP[2][2] = Pold[2][2];

    // (F * P) * F^T + Q
    // F^T = [[1,0,0],[F01,1,0],[F02,F12,1]]
    P(0,0) = FP[0][0] + F01*FP[0][1] + F02*FP[0][2] + Q[0][0];
    P(0,1) = FP[0][1] + F12*FP[0][2] + Q[0][1];
    P(0,2) = FP[0][2]                 + Q[0][2];
    P(1,0) = FP[1][0] + F01*FP[1][1] + F02*FP[1][2] + Q[1][0];
    P(1,1) = FP[1][1] + F12*FP[1][2] + Q[1][1];
    P(1,2) = FP[1][2]                 + Q[1][2];
    P(2,0) = FP[2][0] + F01*FP[2][1] + F02*FP[2][2] + Q[2][0];
    P(2,1) = FP[2][1] + F12*FP[2][2] + Q[2][1];
    P(2,2) = FP[2][2]                 + Q[2][2];
}

// 更新步：卡尔曼增益 + 状态修正 + 协方差更新
// H = [1, 0, 0]（只观测高度）
void VerticalKalmanFilter::update(double zAltitude) {
    // 新息协方差 S = H*P*H^T + R = P[0][0] + R
    double S = P(0, 0) + m_R;
    if (S < 1e-12) return; // 防止除零

    // 卡尔曼增益 K = P*H^T / S = P 第 0 列 / S
    double K0 = P(0, 0) / S;
    double K1 = P(1, 0) / S;
    double K2 = P(2, 0) / S;

    // 状态更新：x = x + K * (z - H*x)
    double y = zAltitude - m_x[0]; // 新息
    m_x[0] += K0 * y;
    m_x[1] += K1 * y;
    m_x[2] += K2 * y;

    // 协方差更新：P = (I - K*H) * P
    // I - K*H = [[1-K0,0,0],[-K1,1,0],[-K2,0,1]]
    double p00 = P(0,0), p01 = P(0,1), p02 = P(0,2);
    double p10 = P(1,0), p11 = P(1,1), p12 = P(1,2);
    double p20 = P(2,0), p21 = P(2,1), p22 = P(2,2);
    P(0,0) = (1-K0)*p00;      P(0,1) = (1-K0)*p01;      P(0,2) = (1-K0)*p02;
    P(1,0) = -K1*p00 + p10;   P(1,1) = -K1*p01 + p11;   P(1,2) = -K1*p02 + p12;
    P(2,0) = -K2*p00 + p20;   P(2,1) = -K2*p01 + p21;   P(2,2) = -K2*p02 + p22;
}

// 零速更新（ZUPT）：强制速度为零，收缩相关协方差
void VerticalKalmanFilter::applyZupt() {
    m_x[1] = 0.0;
    // 收缩第 1 行和第 1 列（速度相关项）
    for (int j = 0; j < 3; ++j) { P(1,j) *= 0.01; P(j,1) *= 0.01; }
    P(1, 1) = 1e-4; // 速度方差设为极小值
}

// ── ZUPTDetector ──────────────────────────────────────────────

// 构造：保存阈值参数
ZUPTDetector::ZUPTDetector(double accThreshold, double gyroThreshold, int windowSize)
    : m_accThreshold(accThreshold)
    , m_gyroThreshold(gyroThreshold)
    , m_windowSize(windowSize)
{}

// 清空历史缓冲区
void ZUPTDetector::reset() {
    m_accBuffer.clear();
    m_gyroBuffer.clear();
}

// 计算滑动窗口方差
double ZUPTDetector::computeVariance(const std::deque<double>& buf) {
    if (buf.empty()) return 0.0;
    double mean = std::accumulate(buf.begin(), buf.end(), 0.0) / buf.size();
    double var  = 0.0;
    for (double v : buf) var += (v - mean) * (v - mean);
    return var / buf.size();
}

// 更新检测器，若窗口满且两个方差均低于阈值则返回 true（静止）
bool ZUPTDetector::update(double accNormSq, double gyroNorm) {
    // 保持滑动窗口大小
    m_accBuffer.push_back(accNormSq);
    m_gyroBuffer.push_back(gyroNorm);
    if ((int)m_accBuffer.size() > m_windowSize) m_accBuffer.pop_front();
    if ((int)m_gyroBuffer.size() > m_windowSize) m_gyroBuffer.pop_front();

    // 窗口未满时不判定
    if ((int)m_accBuffer.size() < m_windowSize) return false;

    double accVar  = computeVariance(m_accBuffer);
    double gyroVar = computeVariance(m_gyroBuffer);
    return (accVar < m_accThreshold) && (gyroVar < m_gyroThreshold);
}
